compiler/src/lnntoamm/Expr.ts

```diff
@@ -565,80 +590,62 @@ class Call extends Expr {
     }
     // first reduction
     const argTys = args.map((arg) => arg.ty);
-    const selFns = FunctionType.matrixSelect(fns, argTys, metadata.scope);
-    fns = selFns.map((selFn) => selFn[0]);
-    const retTy = Type.oneOf(selFns.map(([_fn, ty]) => ty));
+    const retTy = Type.generate();
+    const [selFns, selPTys, selRetTys] = FunctionType.matrixSelect(
+      fns,
+      argTys,
+      retTy,
+      metadata.scope,
+    );
+    fns = selFns;
+    // console.log('initial select for', ast.t.trim(), '->', fns);
+    // stdout.write('-> args ');
+    // console.dir(args, { depth: 4 });
+    args.forEach((arg, ii) =>
+      arg.ty.constrain(Type.oneOf(selPTys[ii]), metadata.scope),
+    );
```

Comment:
Any reason why this isn't operating on the `argTys` array instead?

This seems equivalent and more like the prior code you were writing:

```suggestion
    argTys.forEach((ty, ii) =>
      ty.constrain(Type.oneOf(selPTys[ii]), metadata.scope)
    );
```

Reply:
i simply forgot i had it - fixing